gitwrapper/gitwrapper.1
.\"
.\" Copyright (c) 2023 Stefan Sperling
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt GITWRAPPER 1
.Os
.Sh NAME
.Nm gitwrapper
.Nd invoke an appropriate Git repository server
.Sh SYNOPSIS
.Nm Fl c Sq Cm git-receive-pack Ar repository-path
.Nm Fl c Sq Cm git-upload-pack Ar repository-path
.Sh DESCRIPTION
At one time, the only Git repository server software easily available
was built into
.Xr git-upload-pack 1
and
.Xr git-receive-pack 1
which are part of the
.Xr git 1
suite.
As a result of this, most Git client implementations had the path and
calling conventions expected by
.Xr git 1
compiled in.
.Pp
Times have changed, however.
On a modern system, the administrator may wish to use one of several
available Git repository servers, such as
.Xr gotd 8 .
.Pp
It would be difficult to modify all Git client software typically available
on a system, so most of the authors of alternative Git servers have written
their programs so that they use the same calling conventions as
.Xr git-upload-pack 1
and
.Xr git-receive-pack 1
and may be put into place in their stead.
.Pp
Although having drop-in replacements for
.Xr git-upload-pack 1
and
.Xr git-receive-pack 1
helps in installing alternative Git servers, it essentially makes the
configuration of the system depend on hard installing new programs in
.Pa /usr .
This leads to configuration problems for many administrators, since they may
wish to install a new Git server without altering the system provided
.Pa /usr .
(This may be, for example, to avoid having upgrade problems when a new
version of the system is installed over the old.)
They may also have a shared
.Pa /usr
among several machines, and may wish to avoid placing implicit
configuration information in a read-only
.Pa /usr .
.Pp
The
.Nm
program is designed to replace
.Xr git-upload-pack 1
and
.Xr git-receive-pack 1
and to invoke an appropriate Git server based on configuration information
placed in
.Xr gotd.conf 5 .
This permits the administrator to configure which Git server is to be
invoked on the system at run-time.
Git repositories which are listed in
.Xr gotd.conf 5
and exist on the filesystem will be served by
.Xr gotsh 1 .
Any other Git repositories will be served by
.Xr git-upload-pack 1
and
.Xr git-receive-pack 1
as found in Git's
.Pa libexec
directory, which is
.Pa /usr/local/libexec/git/
by default on
.Ox .
.Sh ENVIRONMENT
.Bl -tag -width GOTD_CONF_PATH
.It Ev GOTD_CONF_PATH
Set the path to the configuration file for
.Xr gotd 8 .
If not specified, the default path
.Pa /etc/gotd.conf
will be used.
.El
.Sh FILES
Configuration for
.Xr gotd 8
is kept in
.Pa /etc/gotd.conf .
.Pp
.Pa git-upload-pack
and
.Pa git-receive-pack
are typically set up as a symlink to
.Nm
which is not usually invoked on its own.
.Sh SEE ALSO
.Xr got 1 ,
.Xr gotd.conf 5 ,
.Xr gotd 8 ,
.Xr mailwrapper 8
.Sh AUTHORS
.An Stefan Sperling Aq Mt stsp@openbsd.org
.Sh BUGS
The entire reason this program exists is a crock.
Instead, a command for invoking a Git server should be standardized or
the Git protocol should be changed to make the path to the program
discoverable by Git clients.


CHANGES
* got 0.87; 2023-04-19
  see git repository history for per-change authorship information
- add gitwrapper(1)
- tog: resume blame and diff search from the first line
- fix crash in got log due to NULL-deref in got_object_blob_close
- add support for protecting references against 'got send -f' to gotd
- fix spurious empty packfile error from gotd when rewinding a branch
- tog: implement automated test harness
- update the base commit ID of unmodified files if the blob ID matches
- fix rebase/histedit -a leaving some files on the temporary branch
- make 'got revert' and 'got rm' work on non-existent directories
- got: flush stdout before printing the error in main()
- when aborting rebase/histedit/merge, unlink files added by merged changes
- fix 'got commit' using a bad parent commit ID when worktree is out-of-date
- allow no-op merge commits to be created
- fix sending merge commits
- show how to fetch a pull request in got.1 pull request example section

* got 0.86; 2023-03-12
  see git repository history for per-change authorship information
- fix race condition on NFS where log-message file's modify time may change
- fix gotd sending too large pack files in some cases
- support histedit fold operations which delete a file and then add it again
- make diffing files which changed into dirs and vice-versa possible
- handle files changing into directories during 'got update'
- add quoting to repository path sent to server for git-shell compatibility
- gotwebd: handle short reads and timeouts
- gotwebd: provide gotweb_render_page() entrypoint for all pages
- gotwebd: reply with non-200 HTTP status code on error

* got 0.85; 2023-03-07
  see git repository history for per-change authorship information
- gotwebd: add missing colon in diff view (patch by Josiah Frentsos)
- more preparation for eventual sha256 object ID support
- add test coverage for more tree conflict cases during merges
- fall back to vi(1) instead of ed(1) if neither EDITOR nor VISUAL are set
- in got.1, clarify what users are expected to do during 'histedit -e'
- gotd requires a config file; don't fail silently when it cannot be read
- regress: replace unportable ln -h option with rm && ln
- regress: make cmdline tests POSIX /bin/sh compatible
- gotd: remove more (all?) double process names in log
- don't pass -d to yacc during the build (patch by Josiah Frentsos)
- regress: override locale settings to force the "C" locale
- regress: replace "sed -i" with ed(1) for portable in-place editing
- fix gotd sometimes reading reused deltas from wrong pack file

* got 0.84; 2023-02-22
  too many changes to list all here; see git repository history for more;
  and see git repository history for per-change authorship information
- add 'got histedit -d' flag to drop all commits
- show worktree UUID in backout/cherrypick -l output
- several changes in preparation for eventual SHA256 object ID support
- make 'got rebase' work when the to-be-rebased branch has no parent commit
- fix bad line-wrapping in tog ref and tree views
- add horizontal scrolling support to tog ref and tree views
- create .pack and .idx files with filemode 0444, like Git does
- make 'got fetch' fetch the work tree's current branch as an implicit fallback
- improve 'got fetch' behaviour when work tree's branch is not on server
- gotwebd: fix briefs/tags navigation overlap
- drop double process name from some gotd logs
- tog: fix high CPU usage issue after starting search (reported by Mikhail)
- fix gotd exiting with abnormal error during client connection teardown
- fix gotd segfault in libevent while disconnecting clients
- tweak 'got commit' log-message validation: now checks timestamp and size > 0
- gotwebd: abort blame if the client disconnects midway through
- make 'got fetch -b branch' only fetch the named branch
- got/tog/gotadmin: call pledge(2) earlier where possible
- no longer error out if redundant "got commit -A $GOT_AUTHOR" option is used
- add 'gotadmin pack' -D flag to force generation of ref-deltas in pack files
- make 'got fetch' update cached HEAD symref if it has changed in remote repo
- add commit/histedit/merge/rebase -C option to commit unresolved conflicts
- make 'got status' check for merge conflict markers on newly added lines only
- fix read/write out of bounds in gitconfig file parser (reported by James Cook)
- fix "got fetch" hanging if remote repo is out-of-date (reported by James Cook)
- ignore patterns with trailing "/" now match directories (suggested by Lucas)
- make 'got merge' honour author setting in gitconfig (reported by James Cook)
- sync default values shown in gotwebd.conf(5) EXAMPLES section
- fix parsing of indented comments in gitconfig file (reported by James Cook)
- fix tog diff between arbitrary commits; regression from 0.80

* got 0.83; 2023-01-30
- fix usage display error in got merge command (patch by Mikhail)
- fix missing commits in pack files created with packed object enumeration
- avoid traversing enumerated commits more than once in got-read-pack
- fix ulimit-related test failures on sparc64 (tracey)
- got commit editor now shows log messages from backout and cherrypick (jamsek)
- new cherrypick/backout -l option to show recorded log messages (jamsek)
- new cherrypick/backout -X option to discard recorded log messages (jamsek)
- gotd: implement the delete-refs capability (op)
- fix histedit -m on a commit which only changes filemode bits
- gotsh.1: show how to set up anonymous public read-only repository access

* got 0.82; 2023-01-23
- fix comparison in tree object parser always evaluating to false (jamsek)
- add missing bounds-check in gitproto ref-line tokenizer
- gotd.8 and gotwebd.8 man page improvements (patch by Josiah Frentsos)
- make gotd session process accept just one flush packet at a time
- sort ENVIRONMENT entries in got.1 and tog.1 (op)
- only forward implicit flush packets from gotsh if they are expected
- return GOT_ERR_EOF from pkt.c if a read attempt indicates EOF

gitwrapper/gitwrapper.c
/*
 * Copyright (c) 2023 Stefan Sperling <stsp@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * Resolve path namespace conflicts for git-upload-pack and git-receive-pack.
 */

#include <sys/queue.h>
#include <sys/types.h>
#include <sys/uio.h>

#include <err.h>
#include <errno.h>
#include <event.h>
#include <imsg.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sha1.h>
#include <sha2.h>
#include <syslog.h>
#include <util.h>
#include <unistd.h>

#include "got_error.h"
#include "got_path.h"

#include "got_lib_dial.h"

#include "gotd.h"
#include "log.h"

#ifndef GITWRAPPER_GIT_LIBEXEC_DIR
#define GITWRAPPER_GIT_LIBEXEC_DIR "/usr/local/libexec/git"
#endif

#ifndef GITWRAPPER_MY_SERVER_PROG
#define GITWRAPPER_MY_SERVER_PROG "gotsh"
#endif


__dead static void
usage(void)
{
	fprintf(stderr, "usage: %s -c '%s|%s repository-path'\n",
	    getprogname(), GOT_DIAL_CMD_SEND, GOT_DIAL_CMD_FETCH);
	exit(1);
}

/*
 * Unveil the specific programs we want to start and hide everything else.
 * This is important to limit the impact of our "exec" pledge.
 */
static const struct got_error *
apply_unveil(const char *myserver)
{
	const char *fetchcmd = GITWRAPPER_GIT_LIBEXEC_DIR "/" \
		GOT_DIAL_CMD_FETCH;
	const char *sendcmd = GITWRAPPER_GIT_LIBEXEC_DIR "/" \
		GOT_DIAL_CMD_SEND;

#ifdef PROFILE
	if (unveil("gmon.out", "rwc") != 0)
		return got_error_from_errno2("unveil", "gmon.out");
#endif
	if (unveil(fetchcmd, "x") != 0)
		return got_error_from_errno2("unveil", fetchcmd);

	if (unveil(sendcmd, "x") != 0)
		return got_error_from_errno2("unveil", sendcmd);

	if (myserver && unveil(myserver, "x") != 0)
		return got_error_from_errno2("unveil", myserver);

	if (unveil(NULL, NULL) != 0)
		return got_error_from_errno("unveil");

	return NULL;
}

int
main(int argc, char *argv[])
{
	const struct got_error *error;
	const char *confpath = NULL;
	char *command = NULL, *repo_name = NULL; /* for matching gotd.conf */
	char *myserver = NULL;
	const char *repo_path = NULL; /* as passed on the command line */
	const char *relpath;
	char *gitcommand = NULL;
	struct gotd gotd;
	struct gotd_repo *repo = NULL;

	log_init(1, LOG_USER); /* Log to stderr. */

#ifndef PROFILE
	if (pledge("stdio rpath exec unveil", NULL) == -1)
		err(1, "pledge");
#endif

	/*
	 * Look up our own server program in PATH so we can unveil(2) it.
	 * This call only errors out upon memory allocation failure.
	 * If the program cannot be found then myserver will be set to NULL.
	 */
	error = got_path_find_prog(&myserver, GITWRAPPER_MY_SERVER_PROG);
	if (error)
		goto done;

	/*
	 * Run parse_config() before unveil(2) because parse_config()
	 * checks whether repository paths exist on disk.
	 * Parsing errors and warnings will be logged to stderr.
	 * Upon failure we will run Git's native tooling so do not
	 * bother checking for errors here.
	 */
	confpath = getenv("GOTD_CONF_PATH");
	if (confpath == NULL)
		confpath = GOTD_CONF_PATH;
	parse_config(confpath, PROC_GOTD, &gotd, 0);

	error = apply_unveil(myserver);
	if (error)
		goto done;

#ifndef PROFILE
	if (pledge("stdio exec", NULL) == -1)
		err(1, "pledge");
#endif

	if (strcmp(getprogname(), GOT_DIAL_CMD_SEND) == 0 ||
	    strcmp(getprogname(), GOT_DIAL_CMD_FETCH) == 0) {
		if (argc != 2)
			usage();
		command = strdup(getprogname());
		if (command == NULL) {
			error = got_error_from_errno("strdup");
			goto done;
		}
		repo_path = argv[1];
		relpath = argv[1];
		while (relpath[0] == '/')
			relpath++;
		repo_name = strdup(relpath);
		if (repo_name == NULL) {
			error = got_error_from_errno("strdup");
			goto done;
		}
	} else {
		if (argc != 3 || strcmp(argv[1], "-c") != 0)
			usage();
		repo_path = argv[2];
		error = got_dial_parse_command(&command, &repo_name,
		    repo_path);
		if (error && error->code == GOT_ERR_BAD_PACKET)
			usage();
		if (error)
			goto done;
	}

	repo = gotd_find_repo_by_name(repo_name, &gotd);

	/*
	 * Invoke our custom Git server if the repository was found
	 * in gotd.conf. Otherwise invoke native git(1) tooling.
	 */
	if (repo) {
		if (myserver == NULL) {
			error = got_error_fmt(GOT_ERR_NO_PROG,
			    "cannot run '%s'",
			    GITWRAPPER_MY_SERVER_PROG);
			goto done;
		}
		execl(myserver, command, repo_name, (char *)NULL);
		error = got_error_from_errno2("execl", myserver);
	} else {
		if (asprintf(&gitcommand, "%s/%s",
		    GITWRAPPER_GIT_LIBEXEC_DIR, command) == -1) {
			error = got_error_from_errno("asprintf");
			goto done;
		}
		execl(gitcommand, gitcommand, repo_path, (char *)NULL);
		error = got_error_from_errno2("execl", gitcommand);
	}

done:
	free(command);
	free(repo_name);
	free(myserver);
	free(gitcommand);
	if (error) {
		fprintf(stderr, "%s: %s\n", getprogname(), error->msg);
		return 1;
	}

	return 0;
}
